app/models/concerns/call_listable.rb

```diff
@@ -8,43 +8,57 @@ module CallListable
   # AND they would otherwise be in the call list
   # (e.g. assigned to current line and in user.patients)
 
-  def recently_called_patients(line)
-    patients.where(line: line)
-            .select { |patient| recently_called_by_user? patient }
+  def ordered_patients(line)
+    call_lists.includes(:patient)
+              .in(line: line)
+              .order_by(order_key: :asc)
+              .map(&:patient)
   end
 
   def call_list_patients(line)
-    patients.where(line: line)
-            .reject { |patient| recently_called_by_user? patient }
+    ordered_patients(line).reject { |x| recently_called_by_user? x }
   end
 
-  def add_patient(patient)
-    patients << patient
-    if call_order
-      reorder_call_list(call_order.unshift(patient.id.to_s))
-    else
-      reorder_call_list([patient.id.to_s])
-    end
-    reload
+  def recently_called_patients(line)
+    ordered_patients(line).select { |x| recently_called_by_user? x }
   end
 
-  def remove_patient(patient)
-    patients.delete patient
+  def recently_reached_patients(line)
+    ordered_patients(line).select { |x| recently_reached_by_user? x }
+  end
+
+  def add_patient(patient)
+    present_calls = call_lists.includes(:patient).where(line: patient.line).to_a
+    return if present_calls.map { |x| x.patient.id.to_s }.include? patient.id.to_s
+    call_lists.where(line: patient.line).destroy_all
     reload
+
+    # Rebuild the call list by putting the new pt in front,
+    # and then recreating the pieces after it
+    call_lists.create! patient: patient,
+                       line: patient.line,
+                       order_key: 0
+    present_calls.each do |entry|
+      call_lists.create! patient: entry.patient,
+                         line: entry.patient.line,
+                         order_key: entry.order_key + 1
+    end
   end
 
-  def reorder_call_list(order)
-    update call_order: order
-    save
+  def remove_patient(patient)
+    call_lists.find_by(patient_id: patient.id).destroy
     reload
+  rescue Mongoid::Errors::DocumentNotFound
   end
 
-  def ordered_patients(line)
-    return call_list_patients(line) unless call_order
-    ordered_patients = call_list_patients(line).sort_by do |patient|
-      call_order.index(patient.id.to_s) || call_order.length
+  def reorder_call_list(order, line)
+    call_lists.destroy_all
```

Comment:
I don't love that is implicitly deletes any call list entries that weren't provided in the new order. Could we make that effect more obvious from the method name, or write this to only drop the reordered entries and shove the rest to the end?